detectable/src/main/java/com/synopsys/integration/detectable/detectables/pnpm/lockfile/process/PnpmYamlTransformer.java

```diff
@@ -235,8 +236,25 @@ private Optional<NameVersion> parseNameVersionFromId(String id) {
     private Optional<Dependency> buildDependencyFromPackageEntry(Map.Entry<String, PnpmPackageInfo> packageEntry) {
         PnpmPackageInfo packageInfo = packageEntry.getValue();
         if (packageInfo.name != null) {
+            // We have the name, this is typical of v6 lock files.
             return Optional.of(Dependency.FACTORY.createNameVersionDependency(Forge.NPMJS, packageInfo.name, packageInfo.version));
         }
+        
+        Optional<NameVersion> optNameVersion = parseNameVersionFromId(packageEntry.getKey());
+        if (packageInfo.version != null && optNameVersion.isPresent()) {
+            // We have the version but not the name. This is common in v9 lock files where custom
+            // content has been added to the project via files or URLs. We'll need to parse the 
+            // name only from the key and then use this specific resolved version the lock file is giving us.
+            // The version in the key will be the name of the URL or file and is useless for our purposes
+            // here.
+            NameVersion nameVersion = optNameVersion.get();
+            
+            return Optional.of(Dependency.FACTORY.createNameVersionDependency(Forge.NPMJS, nameVersion.getName(), packageInfo.version));
+        }
+        
+        // We have neither the name or the version specifically stated in the packageInfo structure.
+        // Attempt to parse them out of the key. This is common in v9 lock files for content in 
+        // public repositories.
         return buildDependencyFromPackageId(packageEntry.getKey());
```

Comment:
What would the different entries in a lockfile look like for the second _**if**_  to be ran versus line 258?

Reply:
So in v9 a typical key will look something like:
'@xmldom/xmldom@0.8.10'

not all packages start with @ but after the last @ there is the version. In this case it is nice, 0.8.10

If something like a file reference or url reference is used you won't get something so clearly defined. Here is a file example:

'@use-gesture/vanilla@file:vanilla.tgz'

So there isn't any parsable version. Thankfully there is a second structure where pnpm has resolved this data:

  '@use-gesture/vanilla@file:vanilla.tgz':
    resolution: {integrity: sha512-2j2V8SiKkZMVz/ft64ulvkerPByDwvdyK5YjY5wsxiNsKiXVio9oibsVwl$
    version: 10.3.1

so we can get the version from there.

Reply:
Looking at the lockfile given in IDETECT-4493, I don't see how we would get the version for vanilla. Is that file incomplete or maybe I am misunderstanding and before we get to the transformer, we have run some build command?

Also, for the https entry:
```
 mssql:
        specifier: https://github.com/tediousjs/node-mssql/archive/refs/tags/v7.1.0.tar.gz
        version: https://github.com/tediousjs/node-mssql/archive/refs/tags/v7.1.0.tar.gz
```
 it looks like we already handled that case (using convertRawEntryToPackageId() ?) and there are no changes in this MR to address that?

Reply:
I created a lockfile using the reproduction steps listed in that issue. Here is the layout I see:

1) At the top of the file there is a importers.dependencies section:

    dependencies:
      '@use-gesture/vanilla':
        specifier: file:vanilla.tgz
        version: file:vanilla.tgz

this version isn't going to be useful to us so we are going to have to look further in the lockfile, specifically in the packages section where we will find:

  '@use-gesture/vanilla@file:vanilla.tgz':
    resolution: {integrity: sha512-2j2V8SiKkZMVz/ft64ulvkerPByDwvdyK5YjY5wsxiNsKiXVio9oibsVwl$
    version: 10.3.1

this section above gets read into the packageInfo variable in the code we are talking about so we can get the version on line 252 by doing packageInfo.version

We have a similar problem with the mssql as you point out. As shown in your comment the version isn't helpful so we again have to look at the packages section where we find:

  mssql@https://github.com/tediousjs/node-mssql/archive/refs/tags/v7.1.0.tar.gz:
    resolution: {tarball: https://github.com/tediousjs/node-mssql/archive/refs/tags/v7.1.0.ta$
    version: 7.1.0
    engines: {node: '>=10'}
    hasBin: true

and can get the 7.1.0 version. So it is the same code that is basically looking at packages section to obtain the version from the more detailed block. Neither file or url worked before this code.

Reply:
I see, thank you for the detailed explanation!